swap/src/protocol/alice/swap.rs

```diff
@@ -160,13 +163,27 @@ async fn next_state(
                     }
                 }
                 enc_sig = event_loop_handle.recv_encrypted_signature() => {
-                    tracing::info!("Received encrypted signature");
+                    let (enc_sig, enc_sig_swap_id) = enc_sig?;
+                    if enc_sig_swap_id == swap_id {
+                        tracing::info!("Received encrypted signature");
 
-                    AliceState::EncSigLearned {
-                        monero_wallet_restore_blockheight,
-                        transfer_proof,
-                        encrypted_signature: Box::new(enc_sig?),
-                        state3,
+                        AliceState::EncSigLearned {
+                            monero_wallet_restore_blockheight,
+                            transfer_proof,
+                            encrypted_signature: Box::new(enc_sig),
+                            state3,
+                        }
+                    } else {
+                        // TODO: Remove unnecessary swap-id check
+                        // This should never happen, because the encsig is sent through the
+                        // channel by swap_id, however, to be save this was included for now
+                        tracing::warn!("Unexpected encsig for swap {} was meant to be for swap {}. \
+                            This encsig will be ignored.", swap_id, enc_sig_swap_id);
+                        AliceState::XmrLockTransferProofSent {
+                            monero_wallet_restore_blockheight,
+                            transfer_proof,
+                            state3,
+                        }
```

Comment:
This is exactly what we shouldn't do (as you say in the comment) and you should be able to avoid that if you associate it the messages with the swap-id when you construct the channels.